### PR TITLE
JS: rename dbscheme predicates to consistently use snake_case in dbscheme

### PR DESCRIPTION
**Pull request by erik-krogh**
Note: This PR is incomplete on purpose, read the end.

The predicates in our `.dbscheme` had about an even mix of snake_case and camelCase names (e.g. `isMethod()` and `parameter_fields()`). 

I've gone through all the JavaScript specific predicates in the `.dbscheme` and renamed them such that we consistently use snake_case.   
This is not a breaking change, as I've put in an upgrade script and added an alias to the old name (in `Aliases.qll` for lack of better place). 

I've also renamed the  `@types` (<- what do we call these?), such that snake_case is consistently used.   
Previously they were a confusing mix of `@snake_case` and `@lowercase`. 

All of this has been done using [a script that I wrote](https://gist.github.com/erik-krogh/25435dfc01843a534508420c1d33807f).   
So in a way you are more reviewing the script than the contents of this PR.   
For this PR I've on purpose limited myself to 10 renames, such that the changes are somewhat readable.   
I'll change this PR to include the full renaming before I undraft it. 
I've tested the full rename (and you can [see the full rename here](https://github.com/erik-krogh/ql/pull/1)), and it works.

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/extractor/ASTExtractor.java [hunk 1]
```diff
@@ -700,7 +700,7 @@ public Label visit(Program nd, Context c) {
           trapwriter.addTuple("isClosureModule", toplevelLabel);
           break;
         case MODULE:
-          trapwriter.addTuple("isES2015Module", toplevelLabel);
+          trapwriter.addTuple("is_ES2015_module", toplevelLabel);
```
Comment (asgerf) [inline javascript/extractor/src/com/semmle/js/extractor/ASTExtractor.java:703]: Could you add a call to `toLowerCase()` so this becomes `is_es2015_module`?

**Review by asgerf — COMMENTED** [review]
This is great! Thanks for going through this, and excellent idea with the script. Makes it a lot easier to review.

A few comments:
- Consistently lower case names (see inline comment)
- Generate QLDoc comments for the aliases:
  - If there's an obvious alternative (not the internal database relation), document what to use instead.
  - Otherwise, maybe just not deprecate it (but still document what it does).

**Comment by erik-krogh** [comment]
@github/codeql-javascript more comments before I rebase this on `main` and put in the full renaming?

**Comment by erik-krogh** [comment]
This is review ready. I got another PR open that should fix the tests.

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 2]
```diff
@@ -826,7 +826,7 @@ class IndexedAccessTypeExpr extends @indexed_access_typeexpr, TypeExpr {
  *
  * In general, there are can more than two operands to an intersection type.
  */
-class IntersectionTypeExpr extends @intersectiontypeexpr, TypeExpr {
+class IntersectionTypeExpr extends @inter_section_typeexpr, TypeExpr {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:829]: Woops, that should be `intersection_typeexpr`

File: javascript/ql/src/semmlecode.javascript.dbscheme [hunk 3]
```diff
@@ -560,7 +560,7 @@ case @typeexpr.kind of
 | 10 = @parenthesized_typeexpr
 | 11 = @tuple_typeexpr
 | 12 = @keyof_typeexpr
-| 13 = @qualifiedtypeaccess
+| 13 = @qualified_typeaccess
```
Comment (asgerf) [inline javascript/ql/src/semmlecode.javascript.dbscheme:563]: Now we have `localtype_access` and `qualified_typeaccess` 😕 

I'd suggest `local_type_access` and `qualified_type_access`.

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 4]
```diff
@@ -1010,7 +1010,7 @@ class VarTypeAccess extends @vartypeaccess, TypeExpr { }
  *
  * This can occur as part of the operand to a `typeof` type or as the first operand to an `is` type.
  */
-class LocalVarTypeAccess extends @localvartypeaccess, VarTypeAccess, LexicalAccess, Identifier {
+class LocalVarTypeAccess extends @localvar_typeaccess, VarTypeAccess, LexicalAccess, Identifier {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:1013]: `local_var_type_access`

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 5]
```diff
@@ -1030,7 +1030,7 @@ class ThisVarTypeAccess extends @thisvartypeaccess, VarTypeAccess { }
  *
  * This can only occur as part of the operand to a `typeof` type.
  */
-class QualifiedVarTypeAccess extends @qualifiedvartypeaccess, VarTypeAccess {
+class QualifiedVarTypeAccess extends @qualifiedvar_typeaccess, VarTypeAccess {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:1033]: `qualified_var_type_access`

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 6]
```diff
@@ -1023,7 +1023,7 @@ class LocalVarTypeAccess extends @localvar_typeaccess, VarTypeAccess, LexicalAcc
  * interface Node { isLeaf(): this is Leaf; }
  * ```
  */
-class ThisVarTypeAccess extends @thisvartypeaccess, VarTypeAccess { }
+class ThisVarTypeAccess extends @thisvar_typeaccess, VarTypeAccess { }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:1026]: `this_var_type_access`

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 7]
```diff
@@ -1318,7 +1318,7 @@ class ImportTypeExpr extends TypeExpr, @importtypeexpr {
   predicate isNamespaceAccess() { this instanceof @import_namespace_access }
 
   /** Holds if this import is used in the context of a variable type, such as `let x: typeof import("fs")`. */
-  predicate isVarTypeAccess() { this instanceof @importvartypeaccess }
+  predicate isVarTypeAccess() { this instanceof @import_var_typeaccess }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:1322]: `import_var_type_access`

File: javascript/ql/src/semmle/javascript/Stmt.qll [hunk 8]
```diff
@@ -67,7 +67,7 @@ class Stmt extends @stmt, ExprOrStmt, Documentable {
 private class TControlStmt =
   TLoopStmt or @if_stmt or @with_stmt or @switch_stmt or @try_stmt or @catchclause;
 
-private class TLoopStmt = TEnhancedForLoop or @while_stmt or @dowhilestmt or @forstmt;
+private class TLoopStmt = TEnhancedForLoop or @while_stmt or @dowhile_stmt or @forstmt;
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Stmt.qll]: `do_while_stmt`

File: javascript/ql/src/semmle/javascript/Stmt.qll [hunk 9]
```diff
@@ -69,7 +69,7 @@ private class TControlStmt =
 
 private class TLoopStmt = TEnhancedForLoop or @while_stmt or @dowhile_stmt or @for_stmt;
 
-private class TEnhancedForLoop = @forinstmt or @foreachstmt or @forofstmt;
+private class TEnhancedForLoop = @forin_stmt or @foreachstmt or @forofstmt;
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Stmt.qll]: `for_in_stmt`

File: javascript/ql/src/semmle/javascript/Stmt.qll [hunk 10]
```diff
@@ -69,7 +69,7 @@ private class TControlStmt =
 
 private class TLoopStmt = TEnhancedForLoop or @while_stmt or @dowhile_stmt or @for_stmt;
 
-private class TEnhancedForLoop = @forin_stmt or @foreachstmt or @forofstmt;
+private class TEnhancedForLoop = @forin_stmt or @foreachstmt or @forof_stmt;
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Stmt.qll]: `for_of_stmt`

File: javascript/ql/src/semmle/javascript/Stmt.qll [hunk 11]
```diff
@@ -69,7 +69,7 @@ private class TControlStmt =
 
 private class TLoopStmt = TEnhancedForLoop or @while_stmt or @dowhile_stmt or @for_stmt;
 
-private class TEnhancedForLoop = @forin_stmt or @foreachstmt or @forof_stmt;
+private class TEnhancedForLoop = @forin_stmt or @foreach_stmt or @forof_stmt;
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Stmt.qll:72]: `for_each_stmt`

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 12]
```diff
@@ -1213,7 +1213,7 @@ class BitNotExpr extends @bit_not_expr, UnaryExpr {
  * typeof A.prototype
  * ```
  */
-class TypeofExpr extends @typeofexpr, UnaryExpr {
+class TypeofExpr extends @type_of_expr, UnaryExpr {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Expr.qll:1216]: `typeof_expr` :trollface:

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 13]
```diff
@@ -2260,7 +2260,7 @@ class ComprehensionBlock extends @comprehensionblock, Expr {
  * ]
  * ```
  */
-class ForInComprehensionBlock extends @forincomprehensionblock, ComprehensionBlock { }
+class ForInComprehensionBlock extends @forin_comprehension_block, ComprehensionBlock { }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Expr.qll:2263]: `for_in_comprehension_block`

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 14]
```diff
@@ -2274,7 +2274,7 @@ class ForInComprehensionBlock extends @forin_comprehension_block, ComprehensionB
  * ]
  * ```
  */
-class ForOfComprehensionBlock extends @forofcomprehensionblock, ComprehensionBlock { }
+class ForOfComprehensionBlock extends @forof_comprehension_block, ComprehensionBlock { }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Expr.qll:2277]: `for_of_comprehension_block`

File: javascript/ql/src/semmle/javascript/JSX.qll [hunk 15]
```diff
@@ -214,7 +214,7 @@ class JSXName extends Expr {
  * { /* TBD *&#47; }
  * </pre>
  */
-class JSXEmptyExpr extends Expr, @jsxemptyexpr { }
+class JSXEmptyExpr extends Expr, @jsxempty_expr { }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/JSX.qll:217]: `jsx_empty_expr`

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 16]
```diff
@@ -1098,7 +1098,7 @@ class ConditionalTypeScope extends @conditionaltypescope, Scope {
  * In the above example, `List` is a concrete expression, `string` is a type annotation,
  * and `List<string>` is thus an expression with type arguments.
  */
-class ExpressionWithTypeArguments extends @expressionwithtypearguments, Expr {
+class ExpressionWithTypeArguments extends @expr_ession_with_type_arguments, Expr {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:1101]: `expression_with_type_arguments`

or the shorter `expr_with_type_arguments` if you prefer

File: javascript/ql/src/semmle/javascript/Variables.qll [hunk 17]
```diff
@@ -94,7 +94,7 @@ class ForScope extends Scope, @for_scope {
 }
 
 /** A scope induced by a `for`-`in` or `for`-`of` statement. */
-class ForInScope extends Scope, @forinscope {
+class ForInScope extends Scope, @forin_scope {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Variables.qll:97]: `for_in_scope`

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 18]
```diff
@@ -2399,7 +2399,7 @@ class ThisType extends Type, @thistype {
  * The type of a named value, `typeof X`, typically denoting the type of
  * a class constructor, namespace object, enum object, or module object.
  */
-class TypeofType extends Type, @typeoftype {
+class TypeofType extends Type, @type_of_type {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/TypeScript.qll:2403]: `typeof_type`

**Review by asgerf — COMMENTED** [review]
I noticed a couple of slip-ups but overall looks great! Do you think you can make the script catch these cases?

Overall the errors seems to be:
- `expr_ession` and `inter_section` being split into two words
- `forof`/`forin` being one word when it should be two
- `type_of` being two words when it's referrring to the `typeof` keyword
- Inconsistent use of `typeaccess` vs `type_access`.

Also, what happened to the relation renaming? I can't seem to find them in the commit history. (If you prefer to reduce the scope to just types, that's fine)

**Comment by erik-krogh** [comment]
> Also, what happened to the relation renaming? I can't seem to find them in the commit history. (If you prefer to reduce the scope to just types, that's fine)

They are there, GitHub just don't handle 250+ commits in a single PR very well. (You see the changes in the [diff for the .dbscheme](https://github.com/github/codeql/pull/4153/files#diff-0709052db8e99e1f51fc94eb38115c07)). 

I'll put the full list of commits here (after I make corrections based on your feedback).

**Review by asgerf — APPROVED** [review]

**Pull request merged**